flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java

```diff
@@ -0,0 +1,158 @@
+ * phase.
+ */
+public class SortMergeReader implements RecordReader<KeyValue> {
+    private final List<RecordReader<KeyValue>> sortedReaders;
+    private final Comparator<RowData> userKeyComparator;
+    private PriorityQueue<Element> minHeap;
+    private Element current;
+
+    public SortMergeReader(
+            List<RecordReader<KeyValue>> sortedReaders, Comparator<RowData> userKeyComparator) {
+        this.sortedReaders = sortedReaders;
+        this.userKeyComparator = userKeyComparator;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        if (sortedReaders.isEmpty()) {
+            return null;
+        }
+        if (Objects.isNull(minHeap)) {
+            heapify();
+        } else {
+            supply();
+        }
+        if (minHeap.size() > 0) {
+            return new SortMergeIterator();
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (RecordReader<KeyValue> reader : sortedReaders) {
+            reader.close();
+        }
+    }
+
+    private void heapify() throws IOException {
+        minHeap =
+                new PriorityQueue<>(
+                        (e1, e2) -> {
+                            int result =
+                                    userKeyComparator.compare(
+                                            e1.iterator.current().key(),
+                                            e2.iterator.current().key());
+                            if (result != 0) {
+                                return result;
+                            }
+                            return Long.compare(
+                                    e1.iterator.current().sequenceNumber(),
+                                    e2.iterator.current().sequenceNumber());
+                        });
+        for (int i = 0; i < sortedReaders.size(); i++) {
+            RecordIterator<KeyValue> iterator = sortedReaders.get(i).readBatch();
+            if (Objects.nonNull(iterator) && iterator.advanceNext()) {
+                minHeap.offer(new Element(iterator, i));
+            }
+        }
+    }
+
+    private void supply() throws IOException {
+        RecordReader<KeyValue> reader = sortedReaders.get(current.index);
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        if (Objects.nonNull(iterator) && iterator.advanceNext()) {
+            minHeap.offer(new Element(iterator, current.index));
+        }
+    }
+
+    /** The iterator iterates on {@link SortMergeReader}. */
+    private class SortMergeIterator implements RecordIterator<KeyValue> {
+        private KeyValue previous;
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            if (minHeap.size() > 0) {
+                current = minHeap.poll();
+                previous = current.iterator.current();
+                if (current.iterator.advanceNext()) {
```

Comment:
I feel that there is a problem here, after `advanceNext`, it is not necessarily the current minimum ah.
(Note the test coverage)

Reply:
Yes, according to the current impl, only previous is ordered. I'll change it.